coot/privateer-wrapper.cc

```diff
@@ -0,0 +1,220 @@
+#include <emscripten/bind.h>
+#include <privateer-lib.h>
+#include <privateer-json.h>
+#include <clipper/clipper-ccp4.h>
+#include <clipper/clipper-minimol.h>
+
+#include <gemmi/mmread.hpp>
+#include "clipper-glyco.h"
+#include "clipper-glyco.cpp"
+#include <regex>
+
+using namespace emscripten;
+struct TorsionEntry { 
+  std::string sugar_1; 
+  std::string sugar_2; 
+  std::string atom_number_1; 
+  std::string atom_number_2; 
+  float phi; 
+  float psi; 
+};
+
+struct TableEntry
+{
+  std::string svg;
+  std::string wurcs;
+  std::string chain;
+  std::string glyconnect_id = "NotFound";
+  std::string glytoucan_id = "NotFound";
+  std::string id; 
+  int torsion_err = 0; 
+  int conformation_err = 0; 
+  int anomer_err = 0; 
+  int puckering_err = 0; 
+  int chirality_err = 0; 
+
+  std::vector<TorsionEntry> torsions;
+};
+
+
+
+std::vector<TableEntry> validate(const std::string &file, const std::string &name)
```

Comment:
This is all good but can we add a function in `molecules_container_js` (see `moorhen-wrappers`) that takes an imol and using the `mmdb::Manager` passes the file contents to this validate function? That way we don't need to get the atoms from JS and pass them back to C++ going through the worker twice. Something similar has been done for `DrawGlycoBlocks` (line 189 at moorhen-wrappers.cc). Here we would need something like this (not tested):
```cpp
        std::vector<TableEntry> privateer_validate(int imol) {
            auto file_contents = molecules_container_t::molecule_to_mmCIF_string(imol);
            return validate(file_contents);
        }
```
Then, in JS there would be no need to have `shim_privateer_validate` and `run_privateer_validation` in the coot worker, and lso no need for `const atoms = await selectedMolecule.getAtoms()` before calling privateer, simply calling `privateer_validate` from the command centre would work:
```js
            const privateerResult = await props.commandCentre.current.cootCommand({
                command: 'privateer_validate',
                commandArgs: [selectedMolecule.molNo],
                returnType: 'privateer_results'
            }, false)
```